internal/function/blame.go

```diff
@@ -0,0 +1,150 @@
+package function
+
+import (
+	"fmt"
+
+	"github.com/src-d/gitbase"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+)
+
+type (
+	// Blame implements git-blame function as UDF
+	Blame struct {
+		repo   sql.Expression
+		commit sql.Expression
+	}
+
+	// BlameLine represents each line of git blame's output
+	BlameLine struct {
+		Commit  string `json:"commit"`
+		File    string `json:"file"`
+		LineNum int    `json:"linenum"`
+		Author  string `json:"author"`
+		Text    string `json:"text"`
+	}
+)
+
+// NewBlame constructor
+func NewBlame(repo, commit sql.Expression) sql.Expression {
+	return &Blame{repo, commit}
+}
+
+func (b *Blame) String() string {
+	return fmt.Sprintf("blame(%s, %s)", b.repo, b.commit)
+}
+
+// Type implements the sql.Expression interface
+func (*Blame) Type() sql.Type {
+	return sql.Array(sql.JSON)
+}
+
+// TransformUp implements the sql.Expression interface
+func (b *Blame) TransformUp(fn sql.TransformExprFunc) (sql.Expression, error) {
+	repo, err := b.repo.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	commit, err := b.commit.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	return fn(&Blame{repo, commit})
+}
+
+// Children implements the Expression interface.
+func (b *Blame) Children() []sql.Expression {
+	return []sql.Expression{b.commit, b.repo}
+}
+
+// IsNullable implements the Expression interface.
+func (b *Blame) IsNullable() bool {
+	return b.repo.IsNullable() || b.commit.IsNullable()
+}
+
+// Resolved implements the Expression interface.
+func (b *Blame) Resolved() bool {
+	return b.commit.Resolved() && b.repo.Resolved()
+}
+
+// Eval implements the sql.Expression interface.
+func (b *Blame) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	span, ctx := ctx.Span("gitbase.Blame")
+	defer span.Finish()
+
+	repo, err := b.resolveRepo(ctx, row)
+	if err != nil {
+		ctx.Warn(0, err.Error())
+		return nil, nil
+	}
+
+	commit, err := b.resolveCommit(ctx, repo, row)
+	if err != nil {
+		ctx.Warn(0, err.Error())
+		return nil, nil
+	}
+
+	fIter, err := commit.Files()
+	if err != nil {
+		ctx.Warn(0, err.Error())
+		return nil, nil
+	}
+	defer fIter.Close()
+
+	var lines []BlameLine
+	for f, err := fIter.Next(); err == nil; f, err = fIter.Next() {
```

Comment:
Instead of computing all blame lines here, you can use `sql.Generator` interface and return that. It's all included in the go-modules branch, so you only have to rebase to it to get those features.

Also, while you're at it, it would be nice to return a struct instead of returning a JSON.

Reply:
I'll look into this, thanks.

Reply:
Done